tests/row.cpp

```diff
@@ -0,0 +1,202 @@
+
+struct mock_value : public std::string {
+    mock_value() = default;
+    mock_value(const char* x) : std::string(x) {}
+};
+
+struct mock_value_2 : public std::string {
+    mock_value_2() = default;
+    mock_value_2(const char* x) : std::string(x) {}
+};
+
+class value_converter_mock {
+public:
+    virtual ~value_converter_mock() {}
+    virtual error_code convert(oid_t oid, const char* bytes, int size,
+        const empty_oid_map&, mock_value& v) = 0;
+    virtual error_code convert_2(oid_t oid, const char* bytes, int size,
+        const empty_oid_map&, mock_value_2& v) = 0;
+};
+
+static value_converter_mock* p_converter_mock;
+
+static error_code convert(oid_t oid, const char* bytes, std::size_t size,
+        const empty_oid_map& oid_map, mock_value& v) {
+    return p_converter_mock->convert(oid, bytes, size, oid_map, v);
+}
+
+static error_code convert(oid_t oid, const char* bytes, std::size_t size,
+        const empty_oid_map& oid_map, mock_value_2& v) {
+    return p_converter_mock->convert_2(oid, bytes, size, oid_map, v);
+}
+
+GTEST("ozo::row") {
+    using ::testing::_;
+    using ::testing::Return;
+    using ::testing::DoAll;
+    using ::testing::SetArgReferee;
+    using ::testing::StrEq;
+    using ::testing::Ref;
+
+    empty_oid_map oid_map{};
+
+    test_row<1> r{ {{{111, "123"}}} };
+    mock_value value;
+
+    ::testing::StrictGMock<value_converter_mock> mock_value_converter;
+    p_converter_mock = ::testing::object(mock_value_converter);
+
+    SHOULD("[perform value conversion only on demand]") {
+        EXPECT_CALL(mock_value_converter, (convert)(111, StrEq("123"), 3, _, Ref(value)))
+            .WillOnce(DoAll(SetArgReferee<4>("321"), Return(error::ok)));
+
+        EXPECT_EQ(error::ok, r.at(0, oid_map, value));
+        EXPECT_EQ(value, "321");
+    }
+
+    SHOULD("[return index_out_of_range if index is out of range]") {
+        EXPECT_EQ(error::row_index_out_of_range, r.at(1, oid_map, value));
+    }
+
+    SHOULD("[return an error if convert_row returns one]") {
+        EXPECT_CALL(mock_value_converter, (convert)(111, StrEq("123"), 3, _, Ref(value)))
+            .WillOnce(Return(error::oid_type_mismatch));
+
+        EXPECT_EQ(error::oid_type_mismatch, r.at(0, oid_map, value));
+        EXPECT_EQ(value, "");
+    }
+
+    SHOULD("[return raw value data on demand]") {
+        std::string_view t;
+        EXPECT_EQ(error::ok, r.text_at(0, t));
+        EXPECT_EQ(t, "123");
+    }
+
+    SHOULD("[fail to return raw data if index is out of range]") {
+        std::string_view t;
+        EXPECT_EQ(error::row_index_out_of_range, r.text_at(1, t));
+    }
+}
+
+struct fusion_adapted {
+    mock_value a;
+    mock_value_2 b;
+};
+
+// TODO: maybe we should try to hide this behind OZO_ADAPT_STRUCT or something?
+// In order to hide usage of fusion or hana as an implementation detail.
+// Would involve a bunch of generated code, though.
+BOOST_FUSION_ADAPT_STRUCT(fusion_adapted, a, b)
```

Comment:
We shouldn't. It is not the best idea to force a user to use yet another one adaptation system for types and structures. The good choice is to support these two systems simultaneously and ask a user only to provide additional information like type name in a target database.

Reply:
ok, comment removed